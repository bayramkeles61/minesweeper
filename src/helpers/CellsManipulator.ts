import { Cell, Coords, Field } from "./Field";

/**
 * Get neighbour cells indexes
 * @param {Coords} coords
 * @returns {Record<string, [number, number]>}
 */
export const getNeighboursItems = ([y, x]: Coords): Record<
  string,
  [number, number]
> => ({
  top: [y - 1, x],
  topRight: [y - 1, x + 1],
  right: [y, x + 1],
  rightBottom: [y + 1, x + 1],
  bottom: [y + 1, x],
  bottomLeft: [y + 1, x - 1],
  left: [y, x - 1],
  leftTop: [y - 1, x - 1],
});

/**
 * Check item in the field
 * @param {Coords} coords
 * @param {Field} field
 * @returns {boolean}
 */
export const checkItemInField = ([y, x]: Coords, { length }: Field): boolean =>
  y >= 0 && x >= 0 && length - y > 0 && length - x > 0;

/**
 * Increment neighbour items for cell with coords
 * @param {Coords} coords
 * @param {Field} field
 * @returns {Cell}
 */
export const incrementNeighbours = (coords: Coords, field: Field): Field => {
  const items = getNeighboursItems(coords);

  for (const [y, x] of Object.values(items)) {
    if (checkItemInField([y, x], field)) {
      const cell = field[y][x];
      if (cell < 8) {
        field[y][x] = (cell + 1) as Cell;
      }
    }
  }

  return field;
};
