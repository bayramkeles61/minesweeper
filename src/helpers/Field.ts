import { incrementNeighbours } from "./CellsManipulator";

export type Cell = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12;
export type Field = Cell[][];
export type Coords = [number, number];

export const CellState: Record<string, Cell> = {
  empty: 0,
  bomb: 9,
  hidden: 10,
  flag: 11,
  weakFlag: 12,
};

/**
 * Create empty field
 *
 * Example emptyFieldGenerator(2)
 * [
 *   [0, 0],
 *   [0, 0],
 * ]
 *
 * @param {number} size
 * @param {Cell} state
 * @returns {Field}
 */
export const generateFieldWithDefaultState = (
  size: number,
  state: Cell = CellState.empty,
): Field => new Array(size).fill(null).map(() => new Array(size).fill(state));

/**
 * Generate field with mines
 * Cell has a range between 0 and 9 where
 * 0 - have no any mine
 * from 0 to 8 - count of mines arrount the cell
 * 9 - special flag for the mine
 *
 * (1/9 ~ 11.11%)
 * Example fieldGenerator(3, 1/9) would return
 * [
 *   [1, 1, 1],
 *   [1, 9, 1],
 *   [1, 1, 1]
 * ]
 *
 * @param {number} size
 * @param {number} probability
 * @returns {Field}
 */
export const fieldGenerator = (size: number, probability: number): Field => {
  if (probability < 0 || probability > 1) {
    throw new Error("Probability must be between 0 and 1");
  }

  let unprocessedCells = size * size;
  let restCellsWithBombs = unprocessedCells * probability;

  const result: Field = generateFieldWithDefaultState(size);

  // Stryker disable next-line EqualityOperator
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      // Stryker disable next-line EqualityOperator
      if (restCellsWithBombs / unprocessedCells > Math.random()) {
        result[y][x] = CellState.bomb;
        incrementNeighbours([y, x], result);
        restCellsWithBombs--;
      }
      unprocessedCells--;
    }
  }
  return result;
};
