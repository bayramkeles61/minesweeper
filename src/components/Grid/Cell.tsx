import React, { FC, memo } from "react";
import styled from "@emotion/styled";

import { Cell as CellType, Coords, CellState } from "@/helpers/Field";

import { useMouseDown } from "@/hooks/useMouseDown";

export interface CellProps {
  /**
   * Cell status based on the CellType
   */
  children: CellType;
  /**
   * Cell coordinates
   */
  coords: Coords;
  /**
   * onClick by cell handler
   */
  onClick: (coords: Coords) => void;
  /**
   * onContextMenu by cell handler
   */
  onContextMenu: (coords: Coords) => void;
}

export const isActiveCell = (cell: CellType): boolean =>
  [CellState.hidden, CellState.flag, CellState.weakFlag].includes(cell);

export const areEqual = (
  prevProps: CellProps,
  nextProps: CellProps,
): boolean => {
  const areEqualCoords =
    prevProps.coords.filter((coord, idx) => nextProps.coords[idx] !== coord)
      .length === 0;

  return (
    prevProps.children === nextProps.children &&
    areEqualCoords &&
    prevProps.onClick === nextProps.onClick &&
    prevProps.onContextMenu === nextProps.onContextMenu
  );
};

export const Cell: FC<CellProps> = memo(({ children, coords, ...rest }) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();

  const onClick = () => rest.onClick(coords);

  const onContextMenu = (event: React.MouseEvent<HTMLDivElement>) => {
    /**
     * Prevent context menu by default
     */
    event.preventDefault();

    if (isActiveCell(children)) {
      rest.onContextMenu(coords);
    }
  };

  const props = {
    onClick,
    onContextMenu,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    mouseDown,
    "data-testid": `${coords}`,
    role: "cell",
  };

  return <ComponentsMap {...props}>{children}</ComponentsMap>;
}, areEqual);

// Stryker disable next-line StringLiteral
Cell.displayName = "Cell";

interface ComponentsMapProps {
  children: CellType;
  onClick: (event: React.MouseEvent<HTMLDivElement>) => void;
  onContextMenu: (event: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown: () => void;
  onMouseUp: () => void;
  onMouseLeave: () => void;
  mouseDown: boolean;
  "data-testid"?: string;
  role: string;
}

const ComponentsMap: FC<ComponentsMapProps> = ({ children, ...rest }) => {
  const nonActiveCellProps = {
    onContextMenu: rest.onContextMenu,
    "data-testid": rest["data-testid"],
    role: rest.role,
  };

  switch (children) {
    case CellState.bomb:
      return (
        <BombFrame {...nonActiveCellProps}>
          <Bomb data-testid={`bomb_${rest["data-testid"]}`} />
        </BombFrame>
      );
    case CellState.hidden:
      return <ClosedFrame {...rest}>{children}</ClosedFrame>;
    case CellState.flag:
      return (
        <ClosedFrame {...rest}>
          <Flag data-testid={`flag_${rest["data-testid"]}`}>{children}</Flag>
        </ClosedFrame>
      );
    case CellState.weakFlag:
      return (
        <ClosedFrame {...rest}>
          <WeakFlag data-testid={`weakFlag_${rest["data-testid"]}`} />
        </ClosedFrame>
      );
    default:
      return <RevealedFrame {...nonActiveCellProps}>{children}</RevealedFrame>;
  }
};

interface ClosedFrameProps {
  mouseDown?: boolean;
}

export const ClosedFrame = styled.div<ClosedFrameProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vh;
  height: 1.8vh;
  color: transparent;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: ${({ mouseDown = false }) =>
    mouseDown ? "transparent" : "white #9e9e9e #9e9e9e white"};
  &:hover {
    filter: brightness(1.1);
  }
`;

const transparent = "rgba(0,0,0,0)";
const colors: { [key in CellType]: string } = {
  0: transparent,
  1: "#2a48ec",
  2: "#2bb13d",
  3: "#ec6561",
  4: "#233db7",
  5: "#a6070f",
  6: "#e400af",
  7: "#906a02",
  8: "#fa0707",
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};

const RevealedFrame = styled(ClosedFrame)`
  border-color: #dddddd;
  cursor: default;
  color: ${({ children }) => colors[children as CellType] ?? transparent};
  &:hover {
    filter: brightness(1);
  }
`;

const Bomb = styled.div`
  border-radius: 50%;
  width: 1vh;
  height: 1vh;
  background-color: #333;
`;

const BombFrame = styled(RevealedFrame)`
  background-color: #ec433c;
`;

const Flag = styled.div`
  width: 0px;
  height: 0px;
  color: ${transparent};
  border-top: 0.5vh solid transparent;
  border-bottom: 0.5vh solid transparent;
  border-left: 0.5vh solid #ec433c;
`;

const WeakFlag = styled(Flag)`
  border-left: 0.5vh solid #f19996;
`;
