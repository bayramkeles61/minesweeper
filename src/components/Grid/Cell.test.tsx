import { describe, it, expect } from "vitest";
import { render, screen, fireEvent, createEvent } from "@testing-library/react";

import { CellState, Cell as CellType, Coords } from "@/helpers/Field";

import { Cell, ClosedFrame, isActiveCell, areEqual } from "./Cell";

describe("Cell component check", () => {
  const coords: Coords = [1, 1];
  const props = {
    coords,
    flagCounter: 0,
    bombs: 10,
    onClick: vi.fn(),
    onContextMenu: vi.fn(),
  };

  for (let cell = CellState.empty; cell <= CellState.weakFlag; cell++) {
    it("Cell renders correct", () => {
      const { asFragment } = render(<Cell {...props}>{cell}</Cell>);

      expect(asFragment()).toMatchSnapshot();
    });
    it("Closed Frame renders correct", () => {
      const { asFragment } = render(<ClosedFrame mouseDown={true} />);
      expect(asFragment()).toMatchSnapshot();
    });
    it("Check prevent default contextMenu for every type of cell", () => {
      render(<Cell {...props}>{cell}</Cell>);

      const cellComp = screen.getByTestId(`${coords}`);

      const contextMenuEvent = createEvent.contextMenu(cellComp);
      fireEvent(cellComp, contextMenuEvent);

      expect(contextMenuEvent.defaultPrevented).toBe(true);
    });

    it("onClick and onContextMenu handler should be called for active cells", () => {
      render(<Cell {...props}>{cell}</Cell>);

      const cellComp = screen.getByTestId(`${coords}`);

      fireEvent.click(cellComp);
      fireEvent.contextMenu(cellComp);

      if (isActiveCell(cell)) {
        expect(props.onClick).toBeCalled();
        expect(props.onContextMenu).toBeCalled();
      } else {
        expect(props.onClick).not.toBeCalled();
        expect(props.onContextMenu).not.toBeCalled();
      }
    });
  }
  it("Check areEqual", () => {
    const prevProps = {
      ...props,
      children: 0 as CellType,
    };

    expect(areEqual(prevProps, { ...prevProps })).toBe(true);

    expect(areEqual(prevProps, { ...prevProps, coords: [2, 1] })).toBe(false);
    expect(areEqual(prevProps, { ...prevProps, coords: [1, 2] })).toBe(false);
    expect(areEqual(prevProps, { ...prevProps, coords: [2, 2] })).toBe(false);
    expect(areEqual(prevProps, { ...prevProps, coords: [1, 0] })).toBe(false);

    expect(areEqual(prevProps, { ...prevProps, children: 1 as CellType })).toBe(
      false,
    );
    expect(areEqual(prevProps, { ...prevProps, onClick: vi.fn() })).toBe(false);
    expect(areEqual(prevProps, { ...prevProps, onContextMenu: vi.fn() })).toBe(
      false,
    );
  });
});
