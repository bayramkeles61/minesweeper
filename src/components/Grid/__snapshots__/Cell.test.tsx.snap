// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cell component check > Cell renders correct 1`] = `
<DocumentFragment>
  <div
    class="css-1l2qare"
    data-testid="1,1"
    role="cell"
  >
    0
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 2`] = `
<DocumentFragment>
  <div
    class="css-y5m8ng"
    data-testid="1,1"
    role="cell"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 3`] = `
<DocumentFragment>
  <div
    class="css-13kl6wr"
    data-testid="1,1"
    role="cell"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 4`] = `
<DocumentFragment>
  <div
    class="css-1ejwm48"
    data-testid="1,1"
    role="cell"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 5`] = `
<DocumentFragment>
  <div
    class="css-bqz2uz"
    data-testid="1,1"
    role="cell"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 6`] = `
<DocumentFragment>
  <div
    class="css-1njursc"
    data-testid="1,1"
    role="cell"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 7`] = `
<DocumentFragment>
  <div
    class="css-1sbs8zf"
    data-testid="1,1"
    role="cell"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 8`] = `
<DocumentFragment>
  <div
    class="css-10ze38e"
    data-testid="1,1"
    role="cell"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 9`] = `
<DocumentFragment>
  <div
    class="css-9s2y9s"
    data-testid="1,1"
    role="cell"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 10`] = `
<DocumentFragment>
  <div
    class="css-12dzs03"
    data-testid="1,1"
    role="cell"
  >
    <div
      class="css-81z0jo"
      data-testid="bomb_1,1"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 11`] = `
<DocumentFragment>
  <div
    class="css-1tt7j02"
    data-testid="1,1"
    role="cell"
  >
    10
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 12`] = `
<DocumentFragment>
  <div
    class="css-1tt7j02"
    data-testid="1,1"
    role="cell"
  >
    <div
      class="css-p26beq"
      data-testid="flag_1,1"
    >
      11
    </div>
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Cell renders correct 13`] = `
<DocumentFragment>
  <div
    class="css-1tt7j02"
    data-testid="1,1"
    role="cell"
  >
    <div
      class="css-15fedo1"
      data-testid="weakFlag_1,1"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 1`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 2`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 3`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 4`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 5`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 6`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 7`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 8`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 9`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 10`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 11`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 12`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;

exports[`Cell component check > Closed Frame renders correct 13`] = `
<DocumentFragment>
  <div
    class="css-8jxsla"
  />
</DocumentFragment>
`;
